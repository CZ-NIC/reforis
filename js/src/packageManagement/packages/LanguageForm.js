/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import { CheckBox } from "foris";

LanguageForm.propTypes = {
    formData: PropTypes.shape({
        languages: PropTypes.arrayOf(PropTypes.shape({
            code: PropTypes.string.isRequired,
            enabled: PropTypes.bool.isRequired,
        })).isRequired,
    }),
    setFormValue: PropTypes.func,
    disabled: PropTypes.bool,
};

export default function LanguageForm({ formData, setFormValue, disabled }) {
    return (
        <>
            <h3>{_("Languages")}</h3>
            <p>{_("If you want to use another language than English you can select it from the following list:")}</p>
            <div className="container mb-3">
                <div className="row justify-content-start">
                    {formData.languages.map((language, idx) => (
                        <div className="col-2 col-lg-1 mr-4" key={language.code}>
                            <CheckBox
                                label={language.code.toUpperCase()}
                                checked={language.enabled}
                                disabled={disabled}

                                onChange={setFormValue((value) => ({
                                    languages: { [idx]: { enabled: { $set: value } } },
                                }))}
                            />
                        </div>
                    ))}
                </div>
            </div>
        </>
    );
}
