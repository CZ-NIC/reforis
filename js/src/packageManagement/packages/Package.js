/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import UserOptions from "./UserOptions/UserOptions";
import PackageCheckBox from "./PackageCheckBox";

Package.propTypes = {
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    enabled: PropTypes.bool.isRequired,
    labels: PropTypes.arrayOf(PropTypes.object).isRequired,
    options: PropTypes.array.isRequired,
    index: PropTypes.number.isRequired,
    setFormValue: PropTypes.func,
    disabled: PropTypes.bool,
};

export default function Package({
    title, description, enabled, labels, options, disabled, setFormValue, index,
}) {
    return (
        <div className="package" key={title}>
            <PackageCheckBox
                name={title}
                labels={labels}
                helpText={description}
                checked={enabled}
                disabled={disabled}

                onChange={setFormValue((value) => ({
                    package_lists: { [index]: { enabled: { $set: value } } },
                }))}
            />
            {options && options.length > 0
                ? (
                    <UserOptions
                        packageIndex={index}
                        options={options}
                        setFormValue={setFormValue}
                        disabled={disabled || !enabled}
                    />
                ) : null}
        </div>
    );
}
