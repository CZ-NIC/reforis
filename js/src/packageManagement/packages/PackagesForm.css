/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

.option {
    margin-left: 3rem;
}
