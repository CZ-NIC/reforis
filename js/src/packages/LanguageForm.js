/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import CheckBox from "common/bootstrap/Checkbox";
import { formFieldsSize } from "common/bootstrap/constants";

LanguageForm.propTypes = {
    formData: PropTypes.shape({
        languages: PropTypes.arrayOf(PropTypes.shape({
            code: PropTypes.string.isRequired,
            enabled: PropTypes.bool.isRequired,
        })).isRequired,
    }),
    setFormValue: PropTypes.func,
    disabled: PropTypes.bool,
};

export default function LanguageForm({ formData, setFormValue, disabled }) {
    return (
        <>
            <h3>{_("Languages")}</h3>
            <p>{_("If you want to use other language than English you can select it from the following list:")}</p>
            <div id="language-packages" className={formFieldsSize}>
                {formData.languages.map((language, idx) => (
                    <CheckBox
                        label={language.code.toUpperCase()}
                        key={idx}
                        checked={language.enabled}
                        useDefaultSize={false}
                        disabled={disabled}

                        onChange={setFormValue((value) => ({
                            languages: { [idx]: { enabled: { $set: value } } },
                        }))}
                    />
                ))}
            </div>
        </>
    );
}
