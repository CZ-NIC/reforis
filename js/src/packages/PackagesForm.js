/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import propTypes from 'prop-types';

import CheckBox from 'common/bootstrap/Checkbox';

PackagesForm.propTypes = {
    formData: propTypes.shape({
        user_lists: propTypes.arrayOf(propTypes.shape({
            title: propTypes.string.isRequired,
            msg: propTypes.string.isRequired,
            enabled: propTypes.bool.isRequired,
        })).isRequired,
    }),
    setFormValue: propTypes.func,
    disabled: propTypes.bool
};

export default function PackagesForm({formData, setFormValue, disabled}) {
    return <>
        <h3>{_('Packages list')}</h3>
        {formData.user_lists.map(
            (_package, idx) => <CheckBox
                label={_package.title}
                key={idx}
                helpText={_package.msg}
                checked={_package.enabled}
                disabled={disabled}

                onChange={setFormValue(value => ({
                    user_lists: {[idx]: {enabled: {$set: value}}}
                }))}
            />
        )}
    </>
}

