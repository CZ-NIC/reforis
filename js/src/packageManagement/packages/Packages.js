/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";
import {
    Alert, ForisURLs, ForisForm, useAPIGet, API_STATE, Spinner,
} from "foris";

import API_URLs from "common/API";
import LanguageForm from "./LanguageForm";
import PackagesForm from "./PackagesForm";

export default function Packages() {
    return (
        <>
            <h1>{_("Packages")}</h1>
            <p>
                {_(`
Apart from the standard installation, you can optionally select bundles of additional software that'd be installed on 
the router. This software can be selected from the following list. Please note that only software that is part of 
TurrisOS or that has been installed from a package list is maintained by automatic updates system. Software that has 
been installed manually or using opkg is not affected.
        `)}
            </p>
            <ForisForm
                forisConfig={{ endpoint: API_URLs.packages }}
                prepData={prepData}
                prepDataToSubmit={prepDataToSubmit}
                validator={validator}
            >
                <DisableIfUpdaterIsDisabled>
                    <PackagesForm />
                    <LanguageForm />
                </DisableIfUpdaterIsDisabled>
            </ForisForm>
        </>
    );
}

DisableIfUpdaterIsDisabled.propTypes = {
    formData: PropTypes.shape({ enabled: PropTypes.bool }),
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]),
};

export function DisableIfUpdaterIsDisabled({ children, formData, ...props }) {
    const [getEnabledState, getEnabled] = useAPIGet(API_URLs.updatesEnabled);

    useEffect(() => {
        getEnabled();
    }, [getEnabled]);

    if (getEnabledState.state !== API_STATE.SUCCESS) {
        return <Spinner />;
    }

    const isDisabled = !getEnabledState.data.enabled;

    const childrenWithFormProps = React.Children.map(
        children,
        (child) => React.cloneElement(child, {
            ...props,
            formData,
            disabled: isDisabled,
        }),
    );

    if (!isDisabled) return childrenWithFormProps;

    return (
        <>
            <DisabledUpdaterAlert />
            <div className="text-muted">
                {childrenWithFormProps}
            </div>
        </>
    );
}

function DisabledUpdaterAlert() {
    const message = _(
        `Please enable <a href="${ForisURLs.packageManagement.updateSettings}">automatic updates</a> to manage packages and languages.`,
    );
    // <Alert /> is used instead of context because warning isn't result of any action on this page
    return (
        <Alert type="warning">
            <span dangerouslySetInnerHTML={{ __html: message }} />
        </Alert>
    );
}

function prepData(formData) {
    formData.package_lists = formData.package_lists
        .filter((_package) => !_package.hidden);
    return formData;
}

function prepDataToSubmit(formData) {
    const packages = formData.package_lists
        .filter((_package) => _package.enabled)
        .map((_package) => {
            const options = _package.options.map((option) => ({
                name: option.name,
                enabled: option.enabled,
            }));
            return {
                name: _package.name,
                options,
            };
        });

    const languages = formData.languages
        .filter((language) => language.enabled)
        .map((language) => language.code);

    return {
        languages,
        package_lists: packages,
    };
}

// Hack to disable submit button
function validator(formData) {
    if (!formData.enabled) return { enabled: true };
    return null;
}
