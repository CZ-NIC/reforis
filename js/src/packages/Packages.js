/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import PropTypes from 'prop-types';

import API_URLs from 'common/API';
import Alert from 'common/bootstrap/Alert';
import {ForisURLs} from 'common/constants';
import ForisForm from 'form/ForisForm';

import LanguageForm from './LanguageForm';
import PackagesForm from './PackagesForm';

export default function Packages() {
    return <>
        <h1>{_('Packages')}</h1>
        <p>{_(`
Apart from the standard installation, you can optionally select bundles of additional software that'd be installed on 
the router. This software can be selected from the following list. Please note that only software that is part of 
TurrisOS or that has been installed from a package list is maintained by automatic updates system. Software that has 
been installed manually or using opkg is not affected.
        `)}</p>
        <ForisForm
            forisConfig={{endpoint: API_URLs.packages}}
            prepData={prepData}
            prepDataToSubmit={prepDataToSubmit}
            validator={validator}
        >
            <DisableIfUpdaterIsDisabled>
                <PackagesForm/>
                <LanguageForm/>
            </DisableIfUpdaterIsDisabled>
        </ForisForm>
    </>
}

DisableIfUpdaterIsDisabled.propTypes = {
    formData: PropTypes.shape({enabled: PropTypes.bool.isRequired}),
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ])
};

export function DisableIfUpdaterIsDisabled({children, ...props}) {
    const isDisabled = !props.formData.enabled;
    const childrenWithFormProps =
        React.Children.map(children, child =>
            React.cloneElement(child, {
                ...props,
                disabled: isDisabled,
            })
        );

    if (!isDisabled)
        return childrenWithFormProps;

    return <>
        <DisabledUpdaterAlert/>
        <div className='text-muted'>
            {childrenWithFormProps}
        </div>
    </>
}

DisabledUpdaterAlert.propTypes = {
    formData: PropTypes.shape({enabled: PropTypes.bool.isRequired})
};

function DisabledUpdaterAlert() {
    const message = _(
        `Please enable <a href="${ForisURLs.updates}">automatic updates</a> to manage packages and languages.`
    );
    return <Alert type='warning'>
        <span dangerouslySetInnerHTML={{__html: message}}/>
    </Alert>
}

function prepData(formData) {
    formData.user_lists = formData.user_lists
        .filter(_package => !_package.hidden);
    return formData
}

function prepDataToSubmit(formData) {
    const packages = formData.user_lists
        .filter(_package => _package.enabled)
        .map(_package => _package.name);

    const languages = formData.languages
        .filter(language => language.enabled)
        .map(language => language.code);

    return {languages: languages, user_lists: packages}
}

// Hack to disable submit button
function validator(formData) {
    if (!formData.enabled)
        return {enabled: true};
    return null;
}
