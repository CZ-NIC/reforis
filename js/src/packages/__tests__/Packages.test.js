/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import {render, queryByText} from 'react-testing-library';

import {packagesFixture} from './__fixtures__/packages';
import Packages from '../Packages';
import mockAxios from 'jest-mock-axios';

describe('<Packages/>', () => {
    let packagesContainer;
    beforeEach(() => {
        const {container} = render(<Packages/>);
        mockAxios.mockResponse({data: packagesFixture()});
        packagesContainer = container;
    });

    it('Test with snapshot.', () => {
        expect(packagesContainer.firstChild).toMatchSnapshot()
    });
    it('Test hidden.', () => {
        const HTMLHiddenPackageMessage = queryByText(packagesContainer, 'Hidden package msg');
        expect(HTMLHiddenPackageMessage).toBeNull();
    })
});
