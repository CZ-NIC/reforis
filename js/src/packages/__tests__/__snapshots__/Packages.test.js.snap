// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Packages/> Updates disabled 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -9,10 +9,26 @@
  TurrisOS or that has been installed from a package list is maintained by automatic updates system. Software that has 
  been installed manually or using opkg is not affected.
          
    </p>
    <form>
+     <div
+       class=\\"alert alert-dismissible alert-warning\\"
+     >
+       <span>
+         Please enable 
+         <a
+           href=\\"/updates\\"
+         >
+           automatic updates
+         </a>
+          to manage packages and languages.
+       </span>
+     </div>
+     <div
+       class=\\"text-muted\\"
+     >
        <h3>
          Packages list
        </h3>
        <div
          class=\\"col-sm-12 offset-lg-1 col-lg-10\\"
@@ -23,10 +39,11 @@
            style=\\"margin-bottom: 0px;\\"
          >
            <input
              checked=\\"\\"
              class=\\"custom-control-input\\"
+             disabled=\\"\\"
              id=\\"1\\"
              type=\\"checkbox\\"
            />
            <label
              class=\\"custom-control-label\\"
@@ -50,10 +67,11 @@
            class=\\"custom-control custom-checkbox\\"
            style=\\"margin-bottom: 0px;\\"
          >
            <input
              class=\\"custom-control-input\\"
+             disabled=\\"\\"
              id=\\"2\\"
              type=\\"checkbox\\"
            />
            <label
              class=\\"custom-control-label\\"
@@ -77,10 +95,11 @@
            class=\\"custom-control custom-checkbox\\"
            style=\\"margin-bottom: 0px;\\"
          >
            <input
              class=\\"custom-control-input\\"
+             disabled=\\"\\"
              id=\\"3\\"
              type=\\"checkbox\\"
            />
            <label
              class=\\"custom-control-label\\"
@@ -104,10 +123,11 @@
            class=\\"custom-control custom-checkbox\\"
            style=\\"margin-bottom: 0px;\\"
          >
            <input
              class=\\"custom-control-input\\"
+             disabled=\\"\\"
              id=\\"4\\"
              type=\\"checkbox\\"
            />
            <label
              class=\\"custom-control-label\\"
@@ -131,10 +151,11 @@
            class=\\"custom-control custom-checkbox\\"
            style=\\"margin-bottom: 0px;\\"
          >
            <input
              class=\\"custom-control-input\\"
+             disabled=\\"\\"
              id=\\"5\\"
              type=\\"checkbox\\"
            />
            <label
              class=\\"custom-control-label\\"
@@ -158,10 +179,11 @@
            class=\\"custom-control custom-checkbox\\"
            style=\\"margin-bottom: 0px;\\"
          >
            <input
              class=\\"custom-control-input\\"
+             disabled=\\"\\"
              id=\\"6\\"
              type=\\"checkbox\\"
            />
            <label
              class=\\"custom-control-label\\"
@@ -185,10 +207,11 @@
            class=\\"custom-control custom-checkbox\\"
            style=\\"margin-bottom: 0px;\\"
          >
            <input
              class=\\"custom-control-input\\"
+             disabled=\\"\\"
              id=\\"7\\"
              type=\\"checkbox\\"
            />
            <label
              class=\\"custom-control-label\\"
@@ -212,10 +235,11 @@
            class=\\"custom-control custom-checkbox\\"
            style=\\"margin-bottom: 0px;\\"
          >
            <input
              class=\\"custom-control-input\\"
+             disabled=\\"\\"
              id=\\"8\\"
              type=\\"checkbox\\"
            />
            <label
              class=\\"custom-control-label\\"
@@ -239,10 +263,11 @@
            class=\\"custom-control custom-checkbox\\"
            style=\\"margin-bottom: 0px;\\"
          >
            <input
              class=\\"custom-control-input\\"
+             disabled=\\"\\"
              id=\\"9\\"
              type=\\"checkbox\\"
            />
            <label
              class=\\"custom-control-label\\"
@@ -266,10 +291,11 @@
            class=\\"custom-control custom-checkbox\\"
            style=\\"margin-bottom: 0px;\\"
          >
            <input
              class=\\"custom-control-input\\"
+             disabled=\\"\\"
              id=\\"10\\"
              type=\\"checkbox\\"
            />
            <label
              class=\\"custom-control-label\\"
@@ -293,10 +319,11 @@
            class=\\"custom-control custom-checkbox\\"
            style=\\"margin-bottom: 0px;\\"
          >
            <input
              class=\\"custom-control-input\\"
+             disabled=\\"\\"
              id=\\"11\\"
              type=\\"checkbox\\"
            />
            <label
              class=\\"custom-control-label\\"
@@ -320,10 +347,11 @@
            class=\\"custom-control custom-checkbox\\"
            style=\\"margin-bottom: 0px;\\"
          >
            <input
              class=\\"custom-control-input\\"
+             disabled=\\"\\"
              id=\\"12\\"
              type=\\"checkbox\\"
            />
            <label
              class=\\"custom-control-label\\"
@@ -347,10 +375,11 @@
            class=\\"custom-control custom-checkbox\\"
            style=\\"margin-bottom: 0px;\\"
          >
            <input
              class=\\"custom-control-input\\"
+             disabled=\\"\\"
              id=\\"13\\"
              type=\\"checkbox\\"
            />
            <label
              class=\\"custom-control-label\\"
@@ -385,10 +414,11 @@
              style=\\"margin-bottom: 0px;\\"
            >
              <input
                checked=\\"\\"
                class=\\"custom-control-input\\"
+               disabled=\\"\\"
                id=\\"14\\"
                type=\\"checkbox\\"
              />
              <label
                class=\\"custom-control-label\\"
@@ -406,10 +436,11 @@
              class=\\"custom-control custom-checkbox\\"
              style=\\"margin-bottom: 0px;\\"
            >
              <input
                class=\\"custom-control-input\\"
+               disabled=\\"\\"
                id=\\"15\\"
                type=\\"checkbox\\"
              />
              <label
                class=\\"custom-control-label\\"
@@ -427,10 +458,11 @@
              class=\\"custom-control custom-checkbox\\"
              style=\\"margin-bottom: 0px;\\"
            >
              <input
                class=\\"custom-control-input\\"
+               disabled=\\"\\"
                id=\\"16\\"
                type=\\"checkbox\\"
              />
              <label
                class=\\"custom-control-label\\"
@@ -448,10 +480,11 @@
              class=\\"custom-control custom-checkbox\\"
              style=\\"margin-bottom: 0px;\\"
            >
              <input
                class=\\"custom-control-input\\"
+               disabled=\\"\\"
                id=\\"17\\"
                type=\\"checkbox\\"
              />
              <label
                class=\\"custom-control-label\\"
@@ -469,10 +502,11 @@
              class=\\"custom-control custom-checkbox\\"
              style=\\"margin-bottom: 0px;\\"
            >
              <input
                class=\\"custom-control-input\\"
+               disabled=\\"\\"
                id=\\"18\\"
                type=\\"checkbox\\"
              />
              <label
                class=\\"custom-control-label\\"
@@ -490,10 +524,11 @@
              class=\\"custom-control custom-checkbox\\"
              style=\\"margin-bottom: 0px;\\"
            >
              <input
                class=\\"custom-control-input\\"
+               disabled=\\"\\"
                id=\\"19\\"
                type=\\"checkbox\\"
              />
              <label
                class=\\"custom-control-label\\"
@@ -511,10 +546,11 @@
              class=\\"custom-control custom-checkbox\\"
              style=\\"margin-bottom: 0px;\\"
            >
              <input
                class=\\"custom-control-input\\"
+               disabled=\\"\\"
                id=\\"20\\"
                type=\\"checkbox\\"
              />
              <label
                class=\\"custom-control-label\\"
@@ -532,10 +568,11 @@
              class=\\"custom-control custom-checkbox\\"
              style=\\"margin-bottom: 0px;\\"
            >
              <input
                class=\\"custom-control-input\\"
+               disabled=\\"\\"
                id=\\"21\\"
                type=\\"checkbox\\"
              />
              <label
                class=\\"custom-control-label\\"
@@ -553,10 +590,11 @@
              class=\\"custom-control custom-checkbox\\"
              style=\\"margin-bottom: 0px;\\"
            >
              <input
                class=\\"custom-control-input\\"
+               disabled=\\"\\"
                id=\\"22\\"
                type=\\"checkbox\\"
              />
              <label
                class=\\"custom-control-label\\"
@@ -574,10 +612,11 @@
              class=\\"custom-control custom-checkbox\\"
              style=\\"margin-bottom: 0px;\\"
            >
              <input
                class=\\"custom-control-input\\"
+               disabled=\\"\\"
                id=\\"23\\"
                type=\\"checkbox\\"
              />
              <label
                class=\\"custom-control-label\\"
@@ -595,24 +634,27 @@
              class=\\"custom-control custom-checkbox\\"
              style=\\"margin-bottom: 0px;\\"
            >
              <input
                class=\\"custom-control-input\\"
+               disabled=\\"\\"
                id=\\"24\\"
                type=\\"checkbox\\"
              />
              <label
                class=\\"custom-control-label\\"
                for=\\"24\\"
              >
                NB
              </label>
+           </div>
          </div>
        </div>
      </div>
      <button
        class=\\"btn btn-primary  offset-lg-8 col-lg-3 col-sm-12\\"
+       disabled=\\"\\"
        type=\\"submit\\"
      >
          Save
      </button>
    </form>"
`;

exports[`<Packages/> Updates enabled 1`] = `
<DocumentFragment>
  <h1>
    Packages
  </h1>
  <p>
    
Apart from the standard installation, you can optionally select bundles of additional software that'd be installed on 
the router. This software can be selected from the following list. Please note that only software that is part of 
TurrisOS or that has been installed from a package list is maintained by automatic updates system. Software that has 
been installed manually or using opkg is not affected.
        
  </p>
  <form>
    <h3>
      Packages list
    </h3>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10"
      style="margin-bottom: 1rem;"
    >
      <div
        class="custom-control custom-checkbox"
        style="margin-bottom: 0px;"
      >
        <input
          checked=""
          class="custom-control-input"
          id="1"
          type="checkbox"
        />
        <label
          class="custom-control-label"
          for="1"
          style="margin-bottom: 0px;"
        >
          Enabled package title
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Enabled package msg
      </small>
    </div>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10"
      style="margin-bottom: 1rem;"
    >
      <div
        class="custom-control custom-checkbox"
        style="margin-bottom: 0px;"
      >
        <input
          class="custom-control-input"
          id="2"
          type="checkbox"
        />
        <label
          class="custom-control-label"
          for="2"
          style="margin-bottom: 0px;"
        >
          Extensions of network protocols for 3G/LTE
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Support for additional protocols and connection types.
      </small>
    </div>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10"
      style="margin-bottom: 1rem;"
    >
      <div
        class="custom-control custom-checkbox"
        style="margin-bottom: 0px;"
      >
        <input
          class="custom-control-input"
          id="3"
          type="checkbox"
        />
        <label
          class="custom-control-label"
          for="3"
          style="margin-bottom: 0px;"
        >
          Device detection
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Software for detecting new devices on local network (EXPERIMENTAL).
      </small>
    </div>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10"
      style="margin-bottom: 1rem;"
    >
      <div
        class="custom-control custom-checkbox"
        style="margin-bottom: 0px;"
      >
        <input
          class="custom-control-input"
          id="4"
          type="checkbox"
        />
        <label
          class="custom-control-label"
          for="4"
          style="margin-bottom: 0px;"
        >
          DVB tuner
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Software for sharing television received by a DVB tuner on Turris. Does not include device drivers.
      </small>
    </div>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10"
      style="margin-bottom: 1rem;"
    >
      <div
        class="custom-control custom-checkbox"
        style="margin-bottom: 0px;"
      >
        <input
          class="custom-control-input"
          id="5"
          type="checkbox"
        />
        <label
          class="custom-control-label"
          for="5"
          style="margin-bottom: 0px;"
        >
          SSH Honeypot
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Trap for password-guessing robots on SSH.
      </small>
    </div>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10"
      style="margin-bottom: 1rem;"
    >
      <div
        class="custom-control custom-checkbox"
        style="margin-bottom: 0px;"
      >
        <input
          class="custom-control-input"
          id="6"
          type="checkbox"
        />
        <label
          class="custom-control-label"
          for="6"
          style="margin-bottom: 0px;"
        >
          LuCI extensions
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Several addional tabs and controls for the advanced LuCI interface.
      </small>
    </div>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10"
      style="margin-bottom: 1rem;"
    >
      <div
        class="custom-control custom-checkbox"
        style="margin-bottom: 0px;"
      >
        <input
          class="custom-control-input"
          id="7"
          type="checkbox"
        />
        <label
          class="custom-control-label"
          for="7"
          style="margin-bottom: 0px;"
        >
          LXC utilities
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Set of utilities to manage Linux Containers (lightweight virtualization technology).
      </small>
    </div>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10"
      style="margin-bottom: 1rem;"
    >
      <div
        class="custom-control custom-checkbox"
        style="margin-bottom: 0px;"
      >
        <input
          class="custom-control-input"
          id="8"
          type="checkbox"
        />
        <label
          class="custom-control-label"
          for="8"
          style="margin-bottom: 0px;"
        >
          NAS
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Services allowing to connect a disk to the router and use it as network data store.
      </small>
    </div>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10"
      style="margin-bottom: 1rem;"
    >
      <div
        class="custom-control custom-checkbox"
        style="margin-bottom: 0px;"
      >
        <input
          class="custom-control-input"
          id="9"
          type="checkbox"
        />
        <label
          class="custom-control-label"
          for="9"
          style="margin-bottom: 0px;"
        >
          Internet connection speed measurement
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Actively measures speed of Internet connection using netmetr.cz service.
      </small>
    </div>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10"
      style="margin-bottom: 1rem;"
    >
      <div
        class="custom-control custom-checkbox"
        style="margin-bottom: 0px;"
      >
        <input
          class="custom-control-input"
          id="10"
          type="checkbox"
        />
        <label
          class="custom-control-label"
          for="10"
          style="margin-bottom: 0px;"
        >
          OpenVPN
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        An easy setup of OpenVPN server from Foris.
      </small>
    </div>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10"
      style="margin-bottom: 1rem;"
    >
      <div
        class="custom-control custom-checkbox"
        style="margin-bottom: 0px;"
      >
        <input
          class="custom-control-input"
          id="11"
          type="checkbox"
        />
        <label
          class="custom-control-label"
          for="11"
          style="margin-bottom: 0px;"
        >
          Pakon
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Software for in depth monitoring of your traffic.
      </small>
    </div>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10"
      style="margin-bottom: 1rem;"
    >
      <div
        class="custom-control custom-checkbox"
        style="margin-bottom: 0px;"
      >
        <input
          class="custom-control-input"
          id="12"
          type="checkbox"
        />
        <label
          class="custom-control-label"
          for="12"
          style="margin-bottom: 0px;"
        >
          Print server
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Services allowing to connect a printer to the router and use it for remote printing.
      </small>
    </div>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10"
      style="margin-bottom: 1rem;"
    >
      <div
        class="custom-control custom-checkbox"
        style="margin-bottom: 0px;"
      >
        <input
          class="custom-control-input"
          id="13"
          type="checkbox"
        />
        <label
          class="custom-control-label"
          for="13"
          style="margin-bottom: 0px;"
        >
          Tor
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Service to increase anonymity on the Internet.
      </small>
    </div>
    <h3>
      Languages
    </h3>
    <p>
      If you want to use other language than English you can select it from the following list:
    </p>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10"
      id="language-packages"
    >
      <div
        class=""
        style="margin-bottom: 1rem;"
      >
        <div
          class="custom-control custom-checkbox"
          style="margin-bottom: 0px;"
        >
          <input
            checked=""
            class="custom-control-input"
            id="14"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="14"
          >
            CS
          </label>
        </div>
      </div>
      <div
        class=""
        style="margin-bottom: 1rem;"
      >
        <div
          class="custom-control custom-checkbox"
          style="margin-bottom: 0px;"
        >
          <input
            class="custom-control-input"
            id="15"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="15"
          >
            DA
          </label>
        </div>
      </div>
      <div
        class=""
        style="margin-bottom: 1rem;"
      >
        <div
          class="custom-control custom-checkbox"
          style="margin-bottom: 0px;"
        >
          <input
            class="custom-control-input"
            id="16"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="16"
          >
            DE
          </label>
        </div>
      </div>
      <div
        class=""
        style="margin-bottom: 1rem;"
      >
        <div
          class="custom-control custom-checkbox"
          style="margin-bottom: 0px;"
        >
          <input
            class="custom-control-input"
            id="17"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="17"
          >
            FR
          </label>
        </div>
      </div>
      <div
        class=""
        style="margin-bottom: 1rem;"
      >
        <div
          class="custom-control custom-checkbox"
          style="margin-bottom: 0px;"
        >
          <input
            class="custom-control-input"
            id="18"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="18"
          >
            LT
          </label>
        </div>
      </div>
      <div
        class=""
        style="margin-bottom: 1rem;"
      >
        <div
          class="custom-control custom-checkbox"
          style="margin-bottom: 0px;"
        >
          <input
            class="custom-control-input"
            id="19"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="19"
          >
            PL
          </label>
        </div>
      </div>
      <div
        class=""
        style="margin-bottom: 1rem;"
      >
        <div
          class="custom-control custom-checkbox"
          style="margin-bottom: 0px;"
        >
          <input
            class="custom-control-input"
            id="20"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="20"
          >
            RU
          </label>
        </div>
      </div>
      <div
        class=""
        style="margin-bottom: 1rem;"
      >
        <div
          class="custom-control custom-checkbox"
          style="margin-bottom: 0px;"
        >
          <input
            class="custom-control-input"
            id="21"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="21"
          >
            SK
          </label>
        </div>
      </div>
      <div
        class=""
        style="margin-bottom: 1rem;"
      >
        <div
          class="custom-control custom-checkbox"
          style="margin-bottom: 0px;"
        >
          <input
            class="custom-control-input"
            id="22"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="22"
          >
            HU
          </label>
        </div>
      </div>
      <div
        class=""
        style="margin-bottom: 1rem;"
      >
        <div
          class="custom-control custom-checkbox"
          style="margin-bottom: 0px;"
        >
          <input
            class="custom-control-input"
            id="23"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="23"
          >
            IT
          </label>
        </div>
      </div>
      <div
        class=""
        style="margin-bottom: 1rem;"
      >
        <div
          class="custom-control custom-checkbox"
          style="margin-bottom: 0px;"
        >
          <input
            class="custom-control-input"
            id="24"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="24"
          >
            NB
          </label>
        </div>
      </div>
    </div>
    <button
      class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
      type="submit"
    >
        Save
    </button>
  </form>
</DocumentFragment>
`;
